airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java

```diff
@@ -89,41 +102,89 @@ public StandardSyncOutput run(StandardSyncInput syncInput, Path jobRoot) throws
               s -> String.format("%s - %s", s.getSyncMode(), s.getDestinationSyncMode()))));
       final StandardTapConfig sourceConfig = WorkerUtils.syncToTapConfig(syncInput);
 
+      final ExecutorService executorService = Executors.newFixedThreadPool(2);
+
       // note: resources are closed in the opposite order in which they are declared. thus source will be
       // closed first (which is what we want).
       try (destination; source) {
         destination.start(destinationConfig, jobRoot);
         source.start(sourceConfig, jobRoot);
 
-        while (!cancelled.get() && !source.isFinished()) {
-          final Optional<AirbyteMessage> maybeMessage = source.attemptRead();
-          if (maybeMessage.isPresent()) {
-            final AirbyteMessage message = mapper.mapMessage(maybeMessage.get());
-
-            messageTracker.accept(message);
-            destination.accept(message);
+        final Thread replicationThread = new Thread(() -> {
+          try {
+            while (!cancelled.get() && !source.isFinished()) {
+              final Optional<AirbyteMessage> messageOptional = source.attemptRead();
+              if (messageOptional.isPresent()) {
+                final AirbyteMessage message = mapper.mapMessage(messageOptional.get());
+
+                LOGGER.info("record in DefaultReplicationWorker: {}", message);
+
+                sourceMessageTracker.accept(message);
+                destination.accept(message);
+              }
+            }
+            destination.notifyEndOfStream();
+          } catch (Exception e) {
+            throw new RuntimeException(e);
           }
-        }
+        });
+
+        final Thread destinationOutputThread = new Thread(() -> {
+          try {
+            while (!cancelled.get() && !destination.isFinished()) {
+              final Optional<AirbyteMessage> messageOptional = destination.attemptRead();
+              if (messageOptional.isPresent()) {
+                LOGGER.info("state in DefaultReplicationWorker from Destination: {}", messageOptional.get());
+                destinationMessageTracker.accept(messageOptional.get());
+              }
+            }
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        });
+
+        final Future<?> destinationThreadExecution = executorService.submit(destinationOutputThread);
+        final Future<?> sourceThreadExecution = executorService.submit(replicationThread);
```

Comment:
these names are confusing. they make it seem like the threads are running the source/dest processes.